File: scripts/Algolia.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'algolia'
require 'dotenv/load'

ALGOLIA_APP_ID = ENV['ALGOLIA_APP_ID']
ALGOLIA_API_KEY = ENV['ALGOLIA_API_KEY']
ALGOLIA_INDEX_NAME = ENV['ALGOLIA_INDEX_NAME']

excludes = %w[notes documentation recovery]
client = Algolia::Search::Client.create(ALGOLIA_APP_ID, ALGOLIA_API_KEY)
index = client.init_index(ALGOLIA_INDEX_NAME)
updates = []

diff = if ARGV[0].eql? '--all'
         Dir['entries/*/*.json']
       else
         `git diff --name-only #{ARGV[1] || 'HEAD^'} entries/`.split("\n")
       end

diff.each do |entry|
  if File.exist? entry
    name, data = JSON.parse(File.read(entry)).first
    puts "Updating #{name}"
    data.merge!({ 'name' => name, 'objectID' => data['domain'] })
    # Rename keys
    data['2fa'] = data.delete 'tfa' if data.key? 'tfa'
    data['category'] = data.delete 'keywords'
    # Remove keys that shouldn't be searchable
    data.reject! { |k, _| excludes.include? k }
    updates.push data
  else
    domain = File.basename(entry, File.extname(entry))
    puts "Removing #{domain}"
    index.delete_object! domain
  end
end

res = index.save_objects(updates)
res.wait

File: scripts/Algolia.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const algoliasearch = require("algoliasearch");
const core = require("@actions/core");
require("dotenv").config(); // Load environment variables from .env file

// Initialize Algolia client and index using environment variables
const client = algoliasearch(
  process.env.ALGOLIA_APP_ID,
  process.env.ALGOLIA_API_KEY
);
const index = client.initIndex(process.env.ALGOLIA_INDEX_NAME);

// Function to process each file
async function processFile(entry) {
  if (fs.existsSync(entry)) {
    // If the file exists, read and parse the JSON content
    const fileContent = JSON.parse(fs.readFileSync(entry, "utf8"));
    const [name, data] = Object.entries(fileContent)[0];
    core.info(`Updating ${name}`);

    // Add name and objectID to the data
    data.name = name;
    data.objectID = data.domain;

    // Rename keys if necessary
    const { tfa, categories, ...rest } = data;
    const renamedData = {
      ...rest,
      ...(tfa && { "2fa": tfa }),
      ...(categories && { category: categories }),
    };

    return renamedData;
  } else {
    // If the file does not exist, remove the corresponding object from Algolia
    const domain = path.basename(entry, ".json");
    core.info(`Removing ${domain}`);
    await index.deleteObject(domain);
    return null;
  }
}

// Main function to process files and update the Algolia index
async function main() {
  const files = process.argv.slice(2); // Get the list of files from command-line arguments
  const updates = [];
  let errors = false;

  // Process each file in parallel
  const results = await Promise.allSettled(
    files.map(async (entry) => processFile(entry))
  );

  results.forEach((result) => {
    if (result.status === "fulfilled" && result.value) {
      updates.push(result.value);
    } else if (result.status === "rejected") {
      errors = true;
      core.error(`Failed to process a file: ${result.reason}`);
    }
  });

  // If there are updates, save them to the Algolia index
  if (updates.length > 0) {
    try {
      await index.saveObjects(updates);
      core.info("All updates saved");
    } catch (err) {
      core.error("Error saving updates:", err);
      errors = true;
    }
  }
  process.exit(+errors);
}

main().catch((err) => {
  core.error(err);
  process.exit(1);
});
